// umi-build-dev/lib/plugins/commands/ui/plugins/blocks/client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dva = require("dva");

var _model = _interopRequireDefault(require("./model"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BlocksViewer = (0, _dva.connect)(function (state) {
  return {
    blocks: state.blocks
  };
})(function (props) {
  function addHandler(name) {
    window.send('blocks', [name]);
  }

  return _react.default.createElement("div", null, _react.default.createElement("h2", null, "Search"), _react.default.createElement("h2", null, "List"), _react.default.createElement("ul", null, props.blocks.data.map(function (item) {
    return _react.default.createElement("li", {
      key: item.name
    }, item.name, _react.default.createElement("button", {
      onClick: addHandler.bind(null, item.name)
    }, "add"));
  })));
});

var _default = function _default(api) {
  api.addPanel({
    title: 'Blocks Viewer',
    path: '/blocks',
    component: BlocksViewer,
    models: [_model.default]
  });
};

exports.default = _default;
