// umi-build-dev/lib/plugins/commands/ui/src/layouts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dva = require("dva");

var _navlink = _interopRequireDefault(require("umi/navlink"));

var _index = _interopRequireDefault(require("./index.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (0, _dva.connect)(function (state) {
  return {
    service: state.service
  };
})(function (props) {
  return _react.default.createElement("div", {
    className: _index.default.normal
  }, _react.default.createElement("div", {
    className: _index.default.header
  }, _react.default.createElement("img", {
    alt: "logo",
    className: _index.default.logo,
    src: "https://gw.alipayobjects.com/zos/rmsportal/lbZMwLpvYYkvMUiqbWfd.png"
  }), "umi ui", _react.default.createElement("sup", null, "alpha")), _react.default.createElement("div", {
    className: _index.default.wrapper
  }, _react.default.createElement("div", {
    className: _index.default.sidebar
  }, _react.default.createElement("ul", null, props.service.panels.map(function (panel, i) {
    return (// eslint-disable-next-line react/no-array-index-key
      _react.default.createElement("li", {
        key: i
      }, _react.default.createElement(_navlink.default, {
        activeClassName: _index.default.active,
        to: panel.path
      }, panel.title))
    );
  }))), _react.default.createElement("div", {
    className: _index.default.main
  }, props.children)));
});

exports.default = _default;
